### PR TITLE
Don't send non-existent extra-lib-dirs to ghc

### LINKED ISSUE #6492: Don't send non-existent extra-lib-dirs to GHC
**Describe the bug**
GHC On Windows is rather strict about `extra-lib-dirs` paths existing. The issue this is causing is that when installing `cabal` through chocolatey I would like to make it as friction less as possible.   Installing `msys2` is not a hard requirement and I can't guarantee the order in which `cabal` and `msys2` are installed.

So I pre-populate the `extra-lib-dirs` with the expected paths to `msys2`. But then this makes it unusable if `msys2` isn't installed at some point since GHC will bail out on the none existing path.

It would be great if we could filter out lib dirs that don't exist when calling GHC. I can't think of a way this can go wrong w.r.t dependencies.

Also see #6304

**To Reproduce**
```
cabal user-config update -a 'extra-lib-dirs: C:\foo'
cabal v2-repl
```

**Expected behavior**

I expect `C:\foo` not to be passed to `ghci` since it doesn't exist.

**System information**
 - Windows 10
 - cabal 3.0.0.0 GHC 8.8.1

**Additional context**
I currently require the user to re-install the cabal package after they install msys2, but that's not very obvious and not very friction free.

### PR DESCRIPTION
**Pull request by gbaz**
Filters extra-lib-dirs before passing to GHC. Resolves #6492 

* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).

### CODE REVIEW

**Comment by Mikolaj** [comment]
@gbaz: is this ready for review?

**Comment by Mikolaj** [comment]
Hey, why is there no label, in that case? :)

**Review by Mikolaj — COMMENTED** [review]
This seems safe (famous last words). Is there a cheap way of protecting this PR from accidental reversion via a test?

**Comment by Ericson2314** [comment]
Should `extra-include-dirs` be handled in the same way for consistency? I don't love making an ad-hoc policy to work around what's arguably a GHC bug, but if we do this everywhere then it's not ad-hoc? :D

**Review by Kleidukos — APPROVED** [review]

**Review by andreasabel — CHANGES_REQUESTED** [review]
I wonder about the placement in `Distribution.Simple.GHC`.  Is this something GHC specific or shouldn't these non-existing directories be filtered out earlier in the pipeline, like after parsing?
Also, I agree with @Ericson2314 's concern that non-existing include directories should also be filtered out.

**Review by andreasabel — APPROVED** [review]
Fine with me.

File: changelog.d/pr-8510 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+synopsis: don't send non-extant extra-lib-dirs to ghc
```
Comment (andreasabel) [inline changelog.d/pr-8510:1]: Didn't know that _non-existent_ can be condensed into _non-extant_ ... Now I know...

**Comment by andreasabel** [comment]
@Mikolaj : I can't remember whether the policy was to (a) squash  the commit with the changelog into one commit with the feature or (b) to keep it separate.  
I guess (a), but feel free to change the label if (b) is the rule.  The contributor guide softly recommends (a): https://github.com/haskell/cabal/blob/7babda92df9e005185e7c9d5e9308a8a96092d7c/CONTRIBUTING.md?plain=1#L240-L242

**Comment by Mikolaj** [comment]
> The contributor guide softly recommends (a)

I think it may be recommending it only because squashing requires one more mental action and users may need the nudge. If commits are well made (each one compiles, they separate refactorings from changes, etc.), as opposed to a string of fix-up commits, IMHO it's beneficial to retain them, including the changelog commit.

**Comment by Mikolaj** [comment]
Ultimately it's up to the user and, normally, the users sets the label.

**Comment by Mikolaj** [comment]
Doh, not user, but dev, PR author.

**Comment by andreasabel** [comment]
> > The contributor guide softly recommends (a)
> 
> I think it may be recommending it only because squashing requires one more mental action and ~users~ devs may need the nudge. If commits are well made (each one compiles, they separate refactorings from changes, etc.), as opposed to a string of fix-up commits, IMHO it's beneficial to retain them, including the changelog commit.

You could add something like this to `CONTRIBUTING.md`.

**Comment by Mikolaj** [comment]
> You could add something like this to CONTRIBUTING.md.

I'd like to keep CONTRIBUTING.md short and not (too much) opinionated. Perhaps just reduce the nudge? Or focus it on the most obvious case?

"If your pull request consists of several fix-up commits (not every commit compiles or they just don't help in understanding the PR, e.g., by separating formatting, refactoring and real functionality changes), please use `squash+merge me` instead of `merge me`..."

**Comment by ulysses4ever** [comment]
I think we should lean to squashing unless the different commits are really different changes. (a) it makes for a clearer history (I doubt anyone benefits from tens or hundreds of commits "add changelog"), (b) individual commits are not tested by the ci, only the overall PR is, isn't it? in that case, there may be unbuildable "middle" commits that will severely complicate any future `git-bisect`ing.

**Comment by Mikolaj** [comment]
Right, different trade-offs and tastes. Given that the PR page retains the original commits, I'd not worry about separating formatting, refactoring and (multiple) changes. However, we may want to migrate off github in the future and I'm can't be sure the PR commits are going to be migrated as well (given that the branch is deleted). Perhaps let's leave it as is. :)

**Pull request merged**